File: Python/programmers/etc/level2/괄호_회전하기.py
def check(s):
    stack = []
    for i in range(len(s)):
        if not stack:
            stack.append(s[i])
        else:
            stack.append(s[i])
            p, q = stack[-2], stack[-1]
            if (p, q) == ('(', ')') or (p, q) == ('{', '}') or (p, q) == ('[', ']'):
                stack.pop()
                stack.pop()
    return len(stack) == 0
        
def solution(s):
    answer = 0
    if check(s):
        answer += 1
    for _ in range(len(s)-1):
        s = s[1:] + s[0]
        if check(s):
            answer += 1
    return answer


s = "[](){}"
print(solution(s)) # 3

s = "}]()[{"
print(solution(s)) # 2

s = "[)(]"
print(solution(s)) # 0

s = "}}}"
print(solution(s)) # 0

File: Java/programmers/etc/level2/괄호회전하기.java
package CodeTest.Java.programmers.etc.level2;

import java.util.*;

public class 괄호회전하기 {
    public int solution(String s) {
        int answer = 0;
        answer += check(s);
        
        for (int i=0; i<s.length()-1; i++) {
            s = s.substring(1) + s.substring(0, 1);
            answer += check(s);
        }
        
        return answer;
    }
    
    private int check(String s) {
        Stack<Character> stk = new Stack<>();
        
        for (Character q : s.toCharArray()) {
            if (stk.isEmpty()) {
                stk.push(q);
            } else {
                char p = stk.peek();
                if ((p=='('&&q==')') || (p=='{'&&q=='}') || (p=='['&&q==']')) stk.pop();
                else stk.push(q);
            }
        }

        return stk.isEmpty() ? 1 : 0;
    }
}
